plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,241 @@
+                return saveQueryGroupInClusterState(queryGroup, currentState);
+            }
+
+            @Override
+            public ThrottlingKey getClusterManagerThrottlingKey() {
+                return createQueryGroupThrottlingKey;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                restoreInflightValues(queryGroup.getResourceLimits());
+                inflightCreateQueryGroupRequestCount.decrementAndGet();
+                logger.warn("failed to save QueryGroup object due to error: {}, for source: {}", e.getMessage(), source);
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                restoreInflightValues(queryGroup.getResourceLimits());
+                inflightCreateQueryGroupRequestCount.decrementAndGet();
+                CreateQueryGroupResponse response = new CreateQueryGroupResponse(queryGroup, RestStatus.OK);
+                listener.onResponse(response);
+            }
+        });
+    }
+
+    /**
+     * Get the allocation value for resourceName for the QueryGroup
+     * @param resourceName - the resourceName we want to get the usage for
+     * @param resourceLimits  - the resource limit from which to get the allocation value for resourceName
+     */
+    private double getResourceLimitValue(String resourceName, final Map<ResourceType, Object> resourceLimits) {
+        for (ResourceType resourceType : resourceLimits.keySet()) {
+            if (resourceType.getName().equals(resourceName)) {
+                return (double) resourceLimits.get(resourceType);
+            }
+        }
+        return 0.0;
+    }
+
+    /**
+     * This method will be executed before we submit the new cluster state
+     * @param queryGroup - the QueryGroup we're currently creating
+     * @param currentClusterState - the cluster state before the update
+     */
+    ClusterState saveQueryGroupInClusterState(final QueryGroup queryGroup, final ClusterState currentClusterState) {
+        final Metadata metadata = currentClusterState.metadata();
+        String groupName = queryGroup.getName();
+        final Map<String, QueryGroup> previousGroups = metadata.queryGroups();
+
+        // check if there's any resource allocation that exceed limit of 1.0
+        String resourceNameWithThresholdExceeded = "";
+        for (ResourceType resourceType : queryGroup.getResourceLimits().keySet()) {
+            String resourceName = resourceType.getName();
+            double existingUsage = calculateExistingUsage(resourceName, previousGroups, groupName);
+            double newGroupUsage = getResourceLimitValue(resourceName, queryGroup.getResourceLimits());
+            inflightResourceLimitValues.computeIfAbsent(resourceName, k -> new DoubleAdder()).add(newGroupUsage);
```

Comment:
Not sure if we need this?  Considering cluster state tasks are processed serially, we should be able to simply rely on (existingUsage + newUsage) and throw exception in case it is breached? cc: @kaushalmahi12

Reply:
That is true!, We might get away with simple primitive data types for this.